include/event_manager/event_manager_priv.h

```diff
@@ -0,0 +1,220 @@
+
+#define _EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio)						\
+		_CONCAT(_CONCAT(event_subscribers_, ename), prio)
+
+#define _EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio)						\
+		STRINGIFY(_EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio))
+
+
+/* Declare a zero-length subscriber. */
+#define _EVENT_SUBSCRIBERS_EMPTY(ename, prio)								\
+	const struct {} _CONCAT(_EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio), empty)			\
+	__attribute__((__section__(STRINGIFY(_EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio))))) = {};
+
+
+/* Convenience macros generating section start and stop markers. */
+
+#define _EVENT_SUBSCRIBERS_START(ename, prio)							\
+		_CONCAT(__start_, _EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio))
+
+#define _EVENT_SUBSCRIBERS_STOP(ename, prio)							\
+		_CONCAT(__stop_,  _EVENT_SUBSCRIBERS_SECTION_PREFIX(ename, prio))
+
+
+#define _EVENT_SUBSCRIBERS_DECLARE(ename)										\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_START(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))[];	\
+	extern const struct event_subscriber _EVENT_SUBSCRIBERS_STOP(ename,  _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))[];
+
+
+/* Macro defining empty subscribers on each priority level.
+ * Each event type keeps an array of subscribers for every priority level.
+ * It can happen that for a given priority no subscriber will be registered.
+ * This macro declares zero-length subscriber that will cause required section
+ * to be generated by the linker. If no subscriber is registered at this
+ * level section will remain empty.
+ */
+#define _EVENT_SUBSCRIBERS_DEFINE(ename)					\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FIRST))	\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_NORMAL))	\
+	_EVENT_SUBSCRIBERS_EMPTY(ename, _SUBS_PRIO_ID(_SUBS_PRIO_FINAL))
+
+
+/* Subscribe a listener to an event. */
+#define _EVENT_SUBSCRIBE(lname, ename, prio)								\
+	const struct event_subscriber _CONCAT(_CONCAT(__event_subscriber_, ename), lname) __used	\
+	__attribute__((__section__(_EVENT_SUBSCRIBERS_SECTION_NAME(ename, prio)))) = {			\
+		.listener = &_CONCAT(__event_listener_, lname),						\
+	}
+
+
+/* Pointer to event type definition is used as event type identifier. */
+#define _EVENT_ID(ename) (&_CONCAT(__event_type_, ename))
+
+
+/* Macro generates a function of name new_ename where ename is provided as
+ * an argument. Allocator function is used to create an event of the given
+ * ename type.
+ */
+#define _EVENT_ALLOCATOR_FN(ename)						\
+	static inline struct ename *_CONCAT(new_, ename)(void)			\
```

Comment:
is dynamic allocation a mandatory thing?

Reply:
How else you would want to do it?
Events are dynamic so you have to do dynamic allocation on them. I expect having a preset of event objects create at build time would be quite limiting.
As for an actual allocator that is used - by default it is heap (k_malloc heap to be precise) but user can hook any other allocator (slab for example).